browser/ipfs/ipfs_dns_resolver_impl.cc

```diff
@@ -0,0 +1,59 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ipfs/ipfs_dns_resolver_impl.h"
+
+#include "chrome/browser/net/secure_dns_config.h"
+#include "chrome/browser/net/system_network_context_manager.h"
+#include "content/public/browser/network_service_instance.h"
+#include "services/network/public/mojom/network_service.mojom.h"
+
+namespace ipfs {
+
+mojo::Remote<network::mojom::DnsConfigChangeManager>
+GetDnsConfigChangeManager() {
+  mojo::Remote<network::mojom::DnsConfigChangeManager>
+      dns_config_change_manager_remote;
+  content::GetNetworkService()->GetDnsConfigChangeManager(
+      dns_config_change_manager_remote.BindNewPipeAndPassReceiver());
+  return dns_config_change_manager_remote;
+}
+
+IpfsDnsResolverImpl::IpfsDnsResolverImpl() {
+  dns_config_change_manager_ = GetDnsConfigChangeManager();
+  dns_config_change_manager_->RequestNotifications(
+      receiver_.BindNewPipeAndPassRemote());
+  receiver_.set_disconnect_handler(base::BindOnce(
+      &IpfsDnsResolverImpl::OnDnsConfigChangeManagerConnectionError,
+      base::Unretained(this)));
+}
+
+IpfsDnsResolverImpl::~IpfsDnsResolverImpl() {}
+
+void IpfsDnsResolverImpl::OnDnsConfigChangeManagerConnectionError() {
+  NOTREACHED() << "DNS Config change manager connection error.";
+}
+
+void IpfsDnsResolverImpl::OnDnsConfigChanged() {
+  Notify(GetDnsOverHttps());
+}
+
+absl::optional<std::string> IpfsDnsResolverImpl::GetDnsOverHttps() {
+  SecureDnsConfig secure_dns_config =
+      SystemNetworkContextManager::GetStubResolverConfigReader()
+          ->GetSecureDnsConfiguration(
+              false /* force_check_parental_controls_for_automatic_mode */);
+  const auto servers = secure_dns_config.doh_servers().servers();
+  if (secure_dns_config.mode() == net::SecureDnsMode::kOff || servers.empty()) {
+    return absl::nullopt;
+  }
+  std::string server_template = servers[0].server_template();
```

Comment:
Why do we only pass the first server to ipfs config? Doesn't DNS.Resolvers config support multiple?

Reply:
https://github.com/ipfs/kubo/issues/8173#issuecomment-1100210145

Reply:
Thanks I understand it now, please rename as GetDnsOverHttpsServer or GetFirstDnsOverHttpsServer as we will always return only the first doh server.